File: python-matrix/neural.py
import math
import numpy as np
import random

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

def relu(x):
    return np.maximum(0, x)

class NeuralNetwork:
    def __init__(self, layers, activation='sigmoid', training='random'):
        self.activation = activation
        self.training = training
        self.layers = []
        for i in range(len(layers) - 1):
            self.layers.append(np.array([ [ random.uniform(-1, 1) for x in range(layers[i + 1]) ] for y in range(layers[i]) ]))

    def run(self, input):
        result = np.array(input)
        for layer in self.layers:
            result = result.dot(layer)
            if self.activation == 'sigmoid':
                result = sigmoid(result)
            if self.activation == 'relu':
                result = relu(result)
        return result

    def likely(self, input):
        result = list(self.run(input))
        maxIndex = result.index(max(result))
        for (symbol, outputs) in self.symbols.items():
            if outputs[maxIndex] == 1:
                return symbol

    def __error(self, trainingItems):
        errorSum = 0
        for item in trainingItems:
            result = self.run(item[0])
            correct = self.symbols[item[1]]
            errorSum += sum([ (result[i] - correct[i]) ** 2 for i in range(len(self.symbols)) ])
        return errorSum / len(trainingItems)

    def train(self, trainingItems, maxError):
        # Setup symbols
        self.symbols = {}
        for item in trainingItems:
            self.symbols[item[1]] = []

        symbols = list(self.symbols.keys())
        for i in range(len(symbols)):
            for j in range(len(symbols)):
                self.symbols[symbols[i]].append(i == j and 1 or 0)

        # Train layers
        trainingCycles = 0
        error = self.__error(trainingItems)
        while error > maxError:
            if self.training == 'random':
                changes = []
                for layer in self.layers:
                    x = random.randint(0, len(layer[0]) - 1)
                    y = random.randint(0, len(layer) - 1)
                    changes.append({ 'layer': layer, 'x': x, 'y': y, 'weight': layer[y][x] })
                    layer[y][x] += random.uniform(-1, 1) / 100

                newError = self.__error(trainingItems)
                if newError < error:
                    error = newError
                else:
                    for change in changes:
                        change['layer'][change['y']][change['x']] = change['weight']

            # I don't know what I'm doing, for more info:
            # http://iamtrask.github.io/2015/07/12/basic-python-network/
            if self.training == 'math':
                inputs = np.array([ item[0] for item in trainingItems ])
                outputs = np.array([ self.symbols[item[1]] for item in trainingItems ])

                results = [ inputs ]
                for layer in self.layers:
                    results.append(sigmoid(results[-1].dot(layer)))

                lastError = None
                for i, layer in reversed(list(enumerate(self.layers))):
                    if i == len(self.layers) - 1:
                        lastError = outputs - results[i + 1]
                        error = np.mean(np.abs(lastError))
                    else:
                        lastError = lastError.dot(self.layers[i + 1].T)

                    layerDelta = lastError * sigmoid_derivative(results[i + 1])
                    layer += results[i].T.dot(layerDelta)

            trainingCycles += 1
        return trainingCycles

File: javascript-matrix/neural.js
export function randint(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

export function sigmoid(x) {
    return 1 / (1 + Math.exp(-x));
}

export function relu(x) {
    return Math.max(0, x);
}

export class Matrix {
    constructor(columns, rows) {
        this.columns = columns;
        this.rows = rows;
        this.elements = new Float32Array(rows * columns);
    }

    get(x, y) {
        return this.elements[y * this.columns + x];
    }

    set(x, y, value) {
        this.elements[y * this.columns + x] = value;
    }

    mul(vector) {
        const result = new Float32Array(this.rows);
        for (let y = 0; y < this.rows; y++) {
            result[y] = 0;
            for (let x = 0; x < this.columns; x++) {
                result[y] += this.elements[y * this.columns + x] * vector[x];
            }
        }
        return result;
    }
}

export class NeuralNetwork {
    constructor({ activation = 'sigmoid', layers }) {
        this.activation = activation;
        this.layers = [];
        for (let i = 0; i < layers.length - 1; i++) {
            const layer = new Matrix(layers[i], layers[i + 1]);
            for (let y = 0; y < layer.rows; y++) {
                for (let x = 0; x < layer.columns; x++) {
                    layer.set(x, y, Math.random() * 2 - 1);
                }
            }
            this.layers.push(layer);
        }
    }

    run(input) {
        let result = new Float32Array(input);
        for (const layer of this.layers) {
            result = layer.mul(result);
            if (this.activation == 'sigmoid') result = result.map(x => sigmoid(x));
            if (this.activation == 'relu') result = result.map(x => relu(x));
        }
        return result;
    }

    likely(input) {
        const result = this.run(input);
        const maxIndex = result.indexOf(Math.max(...result));
        for (const symbol in this.symbols) {
            if (this.symbols[symbol][maxIndex] == 1) {
                return symbol;
            }
        }
    }

    _error(trainingItems) {
        let errorSum = 0;
        for (const item of trainingItems) {
            const result = this.run(item.input);
            const correct = this.symbols[item.output];
            for (let i = 0; i < correct.length; i++) {
                errorSum += (result[i] - correct[i]) ** 2;
            }
        }
        return errorSum / trainingItems.length;
    }

    train(trainingItems, maxError) {
        // Setup symbols
        this.symbols = {};
        for (const item of trainingItems) {
            this.symbols[item.output] = [];
        }
        const symbols = Object.keys(this.symbols);
        for (let i = 0; i < symbols.length; i++) {
            for (let j = 0; j < symbols.length; j++) {
                this.symbols[symbols[i]].push(i == j ? 1 : 0);
            }
        }

        // Train weights
        let trainingCycles = 0;
        let error = this._error(trainingItems);
        while (error > maxError) {
            const changes = [];
            for (const layer of this.layers) {
                const y = randint(0, layer.rows - 1);
                const x = randint(0, layer.columns - 1);
                const weight = layer.get(x, y);
                changes.push({ layer, x, y, weight });
                layer.set(x, y, weight + (Math.random() * 2 - 1) / 100);
            }

            const newError = this._error(trainingItems);
            if (newError < error) {
                error = newError;
            } else {
                for (const change of changes) {
                    change.layer.set(change.x, change.y, change.weight);
                }
            }
            trainingCycles++;
        }
        return trainingCycles;
    }
}
